File: solutions/nim/resistor-color-trio/1/resistor_color_trio.nim
# Dedicated to Shree DR.MDD
import math

type
  ResistorColor * = enum
    Black
    Brown
    Red
    Orange
    Yellow
    Green
    Blue
    Violet
    Grey
    White

proc label*(colors: openArray[ResistorColor]): (int, string) =
  let (d1, d2, d3) = (colors[0].ord, colors[1].ord, colors[2].ord)
  var mantissa = (d1 * 10) + d2
  var factor = 1
  var cnt = 0
  while cnt < d3:
    factor = factor * 10
    inc cnt
  let resultVal = mantissa * factor

  if resultVal >= 1000000000:
    (resultVal div 1000000000, "gigaohms")
  elif resultVal >= 1000000:
    (resultVal div 1000000, "megaohms")
  elif resultVal >= 1000:
    (resultVal div 1000, "kiloohms")
  else:
    (resultVal, "ohms")

File: solutions/lua/resistor-color-trio/1/resistor-color-trio.lua
local COLORS = {
    black = 0, brown = 1, red = 2, orange = 3, yellow = 4,
    green = 5, blue = 6, violet = 7, grey = 8, white = 9
}

local UNITS = { "ohms", "kiloohms", "megaohms", "gigaohms" }

return {
  decode = function(c1, c2, c3)
    local mantissa, exponent = COLORS[c1] + COLORS[c2] * 0.1, COLORS[c3] + 1
    return (mantissa * 10 ^ (exponent % 3)), UNITS[exponent // 3 + 1]
  end
}